# Data Analysis

## Dataset - Facebook metrics Data Set

(Dataset: http://archive.ics.uci.edu/ml/machine-learning-databases/00368/Facebook_metrics.zip)

In [2]:
import pandas as pd

In [16]:
file_name = 'dataset_Facebook.csv'  # dataset filename
# check file existence
try:
    df = pd.read_csv(file_name, delimiter=';')  # read data from csv file
except FileNotFoundError as err:
    print(err)  # output error when there is no file in the "notebook"-file directory

In [21]:
print(f'Rows number: {df.shape[0]}.')
print(f'Columns number: {df.shape[1]}.')

Rows number: 500.
Columns number: 19.


In [24]:
df.head(10)

Page total likes    Type  Category  Post Month  Post Weekday  Post Hour  \
0            139441   Photo         2          12             4          3   
1            139441  Status         2          12             3         10   
2            139441   Photo         3          12             3          3   
3            139441   Photo         2          12             2         10   
4            139441   Photo         2          12             2          3   
5            139441  Status         2          12             1          9   
6            139441   Photo         3          12             1          3   
7            139441   Photo         3          12             7          9   
8            139441  Status         2          12             7          3   
9            139441   Photo         3          12             6         10   

   Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
0   0.0                       2752                             5091   
1   0.0                      10460                            19057   
2   0.0                       2413                             4373   
3   1.0                      50128                            87991   
4   0.0                       7244                            13594   
5   0.0                      10472                            20849   
6   1.0                      11692                            19479   
7   1.0                      13720                            24137   
8   0.0                      11844                            22538   
9   0.0                       4694                             8668   

   Lifetime Engaged Users  Lifetime Post Consumers  \
0                     178                      109   
1                    1457                     1361   
2                     177                      113   
3                    2211                      790   
4                     671                      410   
5                    1191                     1073   
6                     481                      265   
7                     537                      232   
8                    1530                     1407   
9                     280                      183   

   Lifetime Post Consumptions  \
0                         159   
1                        1674   
2                         154   
3                        1119   
4                         580   
5                        1389   
6                         364   
7                         305   
8                        1692   
9                         250   

   Lifetime Post Impressions by people who have liked your Page  \
0                                               3078              
1                                              11710              
2                                               2812              
3                                              61027              
4                                               6228              
5                                              16034              
6                                              15432              
7                                              19728              
8                                              15220              
9                                               4309              

   Lifetime Post reach by people who like your Page  \
0                                              1640   
1                                              6112   
2                                              1503   
3                                             32048   
4                                              3200   
5                                              7852   
6                                              9328   
7                                             11056   
8                                              7912   
9                                              2324   

   Lifetime People who have liked your Page and engaged wit

## Statistics calculation (without "Type" - numerical signs)

In [36]:
# columns required to calculation
req_cols = [col for col in df.head(0) if (col != 'Type')]
indices = ['Mean', 'Max', 'Min', 'Median', 'Mode']
# new dataframe without 'Type' column
description1 = pd.DataFrame(index=indices, columns=req_cols)
# filling new dataframe
for column in req_cols:
    mean = df[column].mean()
    max_value = df[column].max()
    min_value = df[column].min()
    median = df[column].median()
    mode = df[column].mode()[0]
    
    signature = [mean, max_value, min_value, median, mode]
    description1[column] = signature  # adding value to dataframe

In [37]:
description1

Page total likes  Category  Post Month  Post Weekday  Post Hour  \
Mean          123194.176      1.88       7.038          4.15       7.84   
Max           139441.000      3.00      12.000          7.00      23.00   
Min            81370.000      1.00       1.000          1.00       1.00   
Median        129600.000      2.00       7.000          4.00       9.00   
Mode          136393.000      1.00      10.000          7.00       3.00   

            Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
Mean    0.278557                   13903.36                        29585.948   
Max     1.000000                  180480.00                      1110282.000   
Min     0.000000                     238.00                          570.000   
Median  0.000000                    5281.00                         9051.000   
Mode    0.000000                     677.00                         4372.000   

        Lifetime Engaged Users  Lifetime Post Consumers  \
Mean                   920.344                  798.772   
Max                  11452.000                11328.000   
Min                      9.000                    9.000   
Median                 625.500                  551.500   
Mode                   537.000                  182.000   

        Lifetime Post Consumptions  \
Mean                       1415.13   
Max                       19779.00   
Min                           9.00   
Median                      851.00   
Mode                        431.00   

        Lifetime Post Impressions by people who have liked your Page  \
Mean                                            16766.376              
Max                                           1107833.000              
Min                                               567.000              
Median                                           6255.500              
Mode                                             1210.000              

        Lifetime Post reach by people who like your Page  \
Mean                                            6585.488   
Max                                            51456.000   
Min                                              236.000   
Median                                          3417.000   
Mode                                             690.000   

        Lifetime People who have liked your Page and engaged with your post  \
Mean                                              609.986                     
Max                                              4376.000                     
Min                                                 9.000                     
Median                                            412.000                     
Mode                                              403.000                     

        comment         like       share  Total Interactions  
Mean      7.482   177.945892   27.266129              212.12  
Max     372.000  5172.000000  790.000000             6334.00  
Min       0.000     0.000000    0.000000                0.00  
Median    3.000   101.000000   19.000000              123.50  
Mode      0.000    98.000000   13.000000                0.00

## Statistics calculation for "Type" (non-numerical sign)

In [53]:
# new dataframe with only 'Type' column
non_numerical_indices = ['Unique elements', 'The most freqent elements', 
           'The least frequent elements']
description2 = pd.DataFrame(index=non_numerical_indices, columns=['Type'])
the_only_column = 'Type'

num_unique = df[the_only_column].value_counts().count()  # the number of unique elements
most_freq = [df[the_only_column].value_counts().axes[0][i] for i in range(3)]
most_freq = ', '.join(most_freq)  # 3 most frequent elements
least_freq = [df[the_only_column].value_counts().axes[0][-i] for i in range(-3, 0)]
least_freq = ', '.join(least_freq)  # 3 least frequent elements
signature = [num_unique, most_freq, least_freq] 
description2[the_only_column] = signature  # adding value to dataframe

In [54]:
description2

Type
Unique elements                                4
The most freqent elements    Photo, Status, Link
The least frequent elements  Video, Link, Status

## The most popular object in dataset

I decided that the popularity of the object can be estimated as the number_of_likes + number_of_reposts + number_of_comments. Thus, the most popular object is given below.

In [57]:
df[df['Total Interactions'] == df['Total Interactions'].max()]

Page total likes   Type  Category  Post Month  Post Weekday  Post Hour  \
244            130791  Photo         2           7             3          5   

     Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
244   1.0                     180480                           319133   

     Lifetime Engaged Users  Lifetime Post Consumers  \
244                    8072                     4010   

     Lifetime Post Consumptions  \
244                        6242   

     Lifetime Post Impressions by people who have liked your Page  \
244                                             108752              

     Lifetime Post reach by people who like your Page  \
244                                             51456   

     Lifetime People who have liked your Page and engaged with your post  \
244                                               3316                     

     comment    like  share  Total Interactions  
244      372  5172.0  790.0                6334

So, the most popular object is a photo with 5172 likes, 790 shares and 372 comments.